File: implementations/ts-express/src/routes/articles.ts
import { Router } from 'express';

import { requireAuth } from '../middleware/auth';
import { asyncHandler } from '../middleware/asyncHandler';
import { toArticleResponse } from '../mappers';
import { prisma } from '../prisma';
import { ConflictError, NotFoundError } from '../problem';
import {
  articleRequestSchema,
  batchCreateSchema,
  parseOrThrow,
  updateArticleRequestSchema,
} from '../validation/schemas';

/**
 * Article resource routes. Every endpoint below requires a valid bearer token —
 * the guard runs for any request that enters this router.
 */
export const articlesRouter: Router = Router();

articlesRouter.use(requireAuth);

const DEFAULT_PAGE = 1;
const DEFAULT_PAGE_SIZE = 10;
const MAX_PAGE_SIZE = 100;

/** Parses a query value to a positive integer, falling back to a default. */
function toPositiveInt(value: unknown, fallback: number): number {
  if (typeof value !== 'string' || value.trim() === '') {
    return fallback;
  }
  const parsed = Number.parseInt(value, 10);
  return Number.isNaN(parsed) ? fallback : parsed;
}

/** Parses an :id path segment; returns null when it is not a valid integer. */
function parseId(raw: string): number | null {
  if (!/^-?\d+$/.test(raw)) {
    return null;
  }
  const id = Number.parseInt(raw, 10);
  return Number.isNaN(id) ? null : id;
}

// GET /api/articles — filter (name partial/ci, category exact) + pagination.
articlesRouter.get(
  '/api/articles',
  asyncHandler(async (req, res) => {
    const nameFilter = typeof req.query.name === 'string' ? req.query.name : undefined;
    const categoryFilter = typeof req.query.category === 'string' ? req.query.category : undefined;

    const page = Math.max(1, toPositiveInt(req.query.page, DEFAULT_PAGE));
    const pageSize = Math.min(
      Math.max(1, toPositiveInt(req.query.pageSize, DEFAULT_PAGE_SIZE)),
      MAX_PAGE_SIZE,
    );

    const where = {
      ...(nameFilter && nameFilter.trim() !== ''
        ? { name: { contains: nameFilter, mode: 'insensitive' as const } }
        : {}),
      ...(categoryFilter && categoryFilter.trim() !== '' ? { category: categoryFilter } : {}),
    };

    const totalCount = await prisma.article.count({ where });
    const articles = await prisma.article.findMany({
      where,
      orderBy: { articleId: 'asc' },
      skip: (page - 1) * pageSize,
      take: pageSize,
    });

    const totalPages = Math.ceil(totalCount / pageSize);

    res.status(200).json({
      items: articles.map(toArticleResponse),
      page,
      pageSize,
      totalCount,
      totalPages,
      hasPrevious: page > 1,
      hasNext: page < totalPages,
    });
  }),
);

// POST /api/articles — create.
articlesRouter.post(
  '/api/articles',
  asyncHandler(async (req, res) => {
    const input = parseOrThrow(articleRequestSchema, req.body);

    const created = await prisma.article.create({
      data: {
        name: input.name,
        description: input.description,
        category: input.category ?? null,
        price: input.price,
        currency: input.currency ?? null,
      },
    });

    res
      .status(201)
      .location(`/api/articles/${created.articleId}`)
      .json(toArticleResponse(created));
  }),
);

// POST /api/articles-concurrent — batch create (order preserved).
articlesRouter.post(
  '/api/articles-concurrent',
  asyncHandler(async (req, res) => {
    const inputs = parseOrThrow(batchCreateSchema, req.body);

    const created = await Promise.all(
      inputs.map((input) =>
        prisma.article.create({
          data: {
            name: input.name,
            description: input.description,
            category: input.category ?? null,
            price: input.price,
            currency: input.currency ?? null,
          },
        }),
      ),
    );

    res.status(201).json(created.map(toArticleResponse));
  }),
);

// GET /api/articles/:id — fetch one.
articlesRouter.get(
  '/api/articles/:id',
  asyncHandler(async (req, res) => {
    const id = parseId(req.params.id);
    if (id === null) {
      throw new NotFoundError(`Article with ID ${req.params.id} was not found.`);
    }

    const article = await prisma.article.findUnique({ where: { articleId: id } });
    if (!article) {
      throw new NotFoundError(`Article with ID ${id} was not found.`);
    }

    res.status(200).json(toArticleResponse(article));
  }),
);

// PUT /api/articles/:id — replace all fields with optimistic concurrency.
articlesRouter.put(
  '/api/articles/:id',
  asyncHandler(async (req, res) => {
    const id = parseId(req.params.id);
    if (id === null) {
      throw new NotFoundError(`Article with ID ${req.params.id} was not found.`);
    }

    const input = parseOrThrow(updateArticleRequestSchema, req.body);

    // 404 if it does not exist at all; 409 only if it exists but the version is stale.
    const existing = await prisma.article.findUnique({ where: { articleId: id } });
    if (!existing) {
      throw new NotFoundError(`Article with ID ${id} was not found.`);
    }

    // Conditional update on (id, version): matches zero rows when the client's
    // row_version is stale, which we surface as 409 Conflict.
    const result = await prisma.article.updateMany({
      where: { articleId: id, version: input.row_version },
      data: {
        name: input.name,
        description: input.description,
        category: input.category ?? null,
        price: input.price,
        currency: input.currency ?? null,
        version: { increment: 1 },
      },
    });

    if (result.count === 0) {
      throw new ConflictError(
        `Article with ID ${id} was modified by another request. Please retry.`,
      );
    }

    const updated = await prisma.article.findUnique({ where: { articleId: id } });
    res.status(200).json(toArticleResponse(updated!));
  }),
);

// DELETE /api/articles/:id — remove.
articlesRouter.delete(
  '/api/articles/:id',
  asyncHandler(async (req, res) => {
    const id = parseId(req.params.id);
    if (id === null) {
      throw new NotFoundError(`Article with ID ${req.params.id} was not found.`);
    }

    const existing = await prisma.article.findUnique({ where: { articleId: id } });
    if (!existing) {
      throw new NotFoundError(`Article with ID ${id} was not found.`);
    }

    await prisma.article.delete({ where: { articleId: id } });
    res.status(204).end();
  }),
);

File: implementations/python-flask/app/blueprints/articles.py
"""Article CRUD, listing and batch creation. All routes require a JWT bearer."""

from math import ceil

from flask import Blueprint, jsonify, request
from flask_jwt_extended import jwt_required

from ..extensions import db
from ..models import Article
from ..problems import conflict_problem, not_found_problem, validation_problem
from ..validation import validate_create, validate_update

articles_bp = Blueprint("articles", __name__, url_prefix="/api")

MAX_PAGE_SIZE = 100
MAX_BATCH = 100


def _parse_positive_int(raw, default, minimum=1):
    if raw is None:
        return default
    try:
        value = int(raw)
    except (TypeError, ValueError):
        return default
    return value if value >= minimum else minimum


def _new_article(data):
    return Article(
        name=data["name"],
        description=data["description"],
        category=data.get("category"),
        price=data["price"],
        currency=data.get("currency"),
        version=1,
    )


@articles_bp.get("/articles")
@jwt_required()
def list_articles():
    name = request.args.get("name")
    category = request.args.get("category")
    page = _parse_positive_int(request.args.get("page"), default=1)
    page_size = _parse_positive_int(request.args.get("pageSize"), default=10)
    if page_size > MAX_PAGE_SIZE:
        page_size = MAX_PAGE_SIZE  # clamp, never reject

    query = Article.query
    if name:
        query = query.filter(Article.name.ilike(f"%{name}%"))
    if category:
        query = query.filter(Article.category == category)

    total_count = query.count()
    total_pages = ceil(total_count / page_size) if total_count else 0
    items = (
        query.order_by(Article.article_id)
        .offset((page - 1) * page_size)
        .limit(page_size)
        .all()
    )

    return jsonify(
        {
            "items": [a.to_response() for a in items],
            "page": page,
            "pageSize": page_size,
            "totalCount": total_count,
            "totalPages": total_pages,
            "hasPrevious": page > 1,
            "hasNext": page < total_pages,
        }
    ), 200


@articles_bp.post("/articles")
@jwt_required()
def create_article():
    body = request.get_json(silent=True)
    if not isinstance(body, dict):
        return validation_problem({"body": ["Request body must be a JSON object."]})

    errors = {}
    data = validate_create(body, errors)
    if errors:
        return validation_problem(errors)

    article = _new_article(data)
    db.session.add(article)
    db.session.commit()

    response = jsonify(article.to_response())
    response.status_code = 201
    response.headers["Location"] = f"/api/articles/{article.article_id}"
    return response


@articles_bp.get("/articles/<int:article_id>")
@jwt_required()
def get_article(article_id):
    article = db.session.get(Article, article_id)
    if article is None:
        return not_found_problem(f"Article with id {article_id} was not found.")
    return jsonify(article.to_response()), 200


@articles_bp.put("/articles/<int:article_id>")
@jwt_required()
def update_article(article_id):
    body = request.get_json(silent=True)
    if not isinstance(body, dict):
        return validation_problem({"body": ["Request body must be a JSON object."]})

    errors = {}
    data = validate_update(body, errors)
    if errors:
        return validation_problem(errors)

    article = db.session.get(Article, article_id)
    if article is None:
        return not_found_problem(f"Article with id {article_id} was not found.")

    if article.version != data["row_version"]:
        return conflict_problem(
            f"Article with id {article_id} was modified by another request. Please retry."
        )

    article.name = data["name"]
    article.description = data["description"]
    article.category = data.get("category")
    article.price = data["price"]
    article.currency = data.get("currency")
    article.version += 1
    db.session.commit()

    return jsonify(article.to_response()), 200


@articles_bp.delete("/articles/<int:article_id>")
@jwt_required()
def delete_article(article_id):
    article = db.session.get(Article, article_id)
    if article is None:
        return not_found_problem(f"Article with id {article_id} was not found.")
    db.session.delete(article)
    db.session.commit()
    return "", 204


@articles_bp.post("/articles-concurrent")
@jwt_required()
def batch_create_articles():
    body = request.get_json(silent=True)
    if not isinstance(body, list):
        return validation_problem({"body": ["Request body must be a JSON array."]})
    if len(body) == 0:
        return validation_problem({"items": ["At least one article is required."]})
    if len(body) > MAX_BATCH:
        return validation_problem(
            {"items": [f"A maximum of {MAX_BATCH} articles can be created at once."]}
        )

    errors = {}
    cleaned = []
    for index, item in enumerate(body):
        if not isinstance(item, dict):
            errors[f"items[{index}]"] = ["Each item must be a JSON object."]
            continue
        item_errors = {}
        data = validate_create(item, item_errors)
        if item_errors:
            for field, messages in item_errors.items():
                errors[f"items[{index}].{field}"] = messages
        else:
            cleaned.append(data)

    if errors:
        return validation_problem(errors)

    articles = [_new_article(data) for data in cleaned]
    db.session.add_all(articles)
    db.session.commit()

    return jsonify([a.to_response() for a in articles]), 201
